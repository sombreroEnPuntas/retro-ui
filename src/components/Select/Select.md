Select

```jsx
<Select name="select">
  <option label="<">{'back'}</option>
  <option label="X">{'stop'}</option>
  <option label=">">{'next'}</option>
</Select>
```

```jsx
<Select name="select">
  <option></option>
  <option>{'\\'}</option>
  <option>{'|'}</option>
  <option>
    {
      '//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////'
    }
  </option>
</Select>
```

```jsx
<Select multiple name="select multiple">
  <option></option>
  <option>{'\\'}</option>
  <option>{'|'}</option>
  <option>
    {
      '//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////'
    }
  </option>
</Select>
```

Select disabled

```jsx
<Select name="select disabled" status="disabled">
  <option></option>
  <option>{'\\'}</option>
  <option>{'|'}</option>
  <option>{'/'}</option>
</Select>
```

Select error

```jsx
<Select name="select error" status="error">
  <option></option>
  <option>{'\\'}</option>
  <option>{'|'}</option>
  <option>{'/'}</option>
</Select>
```

Select success

```jsx
<Select name="select success" status="success">
  <option></option>
  <option>{'\\'}</option>
  <option>{'|'}</option>
  <option>{'/'}</option>
</Select>
```
