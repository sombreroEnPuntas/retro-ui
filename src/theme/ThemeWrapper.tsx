import React from 'react'
import { ThemeProvider } from 'styled-components'

import { theme, Theme } from './theme'

export const ThemeWrapper: React.FunctionComponent<{ customTheme?: Theme }> = ({
  children,
  customTheme,
}) => (
  <ThemeProvider theme={{ ...theme, ...customTheme }}>{children}</ThemeProvider>
)

export default ThemeWrapper
