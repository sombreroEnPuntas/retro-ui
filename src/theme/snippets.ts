export const boxShadow = ({
  colors,
  type,
}: {
  colors: {
    [key: string]: {
      shadow: string
    }
  }
  type: string
}) => `
  box-shadow: \
    0 4px ${colors[type].shadow}, \
    0 -4px ${colors[type].shadow}, \
    4px 0 ${colors[type].shadow}, \
    -4px 0 ${colors[type].shadow};
  margin: 4px;
`
