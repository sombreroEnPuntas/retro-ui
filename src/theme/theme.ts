import baseStyled, { keyframes, ThemedStyledInterface } from 'styled-components'

const animations = {
  blink: keyframes`
  0% {
    opacity: .2;
  }
  20% {
    opacity: 1;
  }
  100% {
    opacity: .2;
  }
  `,
}

const colors = {
  black: {
    background: 'white',
    backgroundAlt: 'papayawhip',
    outline: '#212529',
    shadow: '#4d5256',
  },
  disabled: '#ebebe4',
  error: {
    background: '#ffd7cf',
    outline: '#ce372b',
    shadow: '#e76e55',
  },
  success: {
    background: '#c9d68f',
    outline: '#76c442',
    shadow: '#92cc41',
  },
  warning: {
    background: '#FFE5A6',
    outline: '#B38106',
    shadow: '#E9C46A',
  },
}

const fonts = {
  fontFamily: "'Press Start 2P', monospace",
  lineHeight: '1.5em',
}

export const theme = {
  animations,
  colors,
  fonts,
}

export type Theme = typeof theme
export const styled = baseStyled as ThemedStyledInterface<Theme>
