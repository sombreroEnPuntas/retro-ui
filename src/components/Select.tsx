import React, { ComponentProps, HTMLProps } from 'react'
import { styled, Theme } from '../theme/theme'

interface LabelProps extends HTMLProps<HTMLLabelElement> {
  multiple?: boolean
  type?: keyof Theme['colors']
}
const Label = styled.label<LabelProps>`
  ${({ theme: { colors, breakpoints, fonts }, multiple, type = 'black' }) => `
    font-family: ${fonts.fontFamily};

    &::after {
      box-shadow: 3px 3px, 6px 3px, 9px 3px, 12px 3px, 15px 3px, 6px 6px, 9px 6px,
        12px 6px, 9px 9px;
      color: ${multiple ? 'transparent' : colors[type].outline};
      content: '';
      display: block;
      height: 3px;
      pointer-events: none;
      position: relative;
      top: calc(-2.2 * ${fonts.lineHeight});
      left: calc(100% - 23px);
      width: 3px;

      ${breakpoints['desktop']} {
        top: calc(-1.5 * ${fonts.lineHeight});
      }
    }
  `}
`

interface SelectProps extends HTMLProps<HTMLSelectElement> {
  status?: keyof Theme['colors'] | 'disabled'
}
const Select = styled.select.attrs(({ status }: SelectProps) => ({
  disabled: status === 'disabled',
}))<SelectProps>`
  ${({ theme: { breakpoints, colors, fonts, snippets }, status = 'black' }) => {
    const type = status === 'disabled' ? 'warning' : status
    return `
      appearance: none;
      background-color: ${
        colors[status === 'disabled' ? 'warning' : 'black'].background
      };
      border: none;
      font-family: ${fonts.fontFamily};
      font-size: ${fonts.lineHeight};
      line-height: calc(2 * ${fonts.lineHeight});
      outline-color: ${colors[type].outline};
      overflow: hidden;
      padding: 0.5rem 1rem;
      text-overflow: ellipsis;
      // https://caniuse.com/#feat=mdn-css_properties_width_stretch
      width: stretch;

      /* Hide arrow icon in IE browsers */
      &::-ms-expand {
        display: none;
      }

      ${snippets.boxShadow({ colors, type })}

      ${breakpoints['desktop']} {
        line-height: ${fonts.lineHeight};
      }

      & option {
        background-color: ${colors.black.background};
        color: ${colors.black.outline};
        overflow: hidden;
        text-overflow: ellipsis;
      }
      & option:checked {
        background-color: ${colors.black.backgroundAlt};
        color: ${colors.black.outline};
      }
    `
  }}
`

const Wrapper = (props: ComponentProps<typeof Select>) => (
  <Label
    multiple={props.multiple}
    type={props.status === 'disabled' ? 'warning' : props.status}
  >
    {props.name}
    <Select {...props} />
  </Label>
)

/** @component */
export default Wrapper
