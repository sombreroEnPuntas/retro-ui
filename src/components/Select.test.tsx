import React from 'react'
import { render } from '@testing-library/react'

// Dependencies
import { ThemeWrapper } from '../theme/ThemeWrapper'

// Unit
import Select from './Select'

const getProps = (customProps: any) => ({
  'data-testid': 'select',
  name: 'test',
  children: [
    <option key="<">{'<'}</option>,
    <option key="X">{'X'}</option>,
    <option key=">">{'>'}</option>,
  ],
  ...customProps,
})

describe.each`
  displayName          | props                     | css
  ${'Select'}          | ${{}}                     | ${'box-shadow: 0 4px #4d5256,0 -4px #4d5256,4px 0 #4d5256,-4px 0 #4d5256; outline-color: #212529;'}
  ${'Select multiple'} | ${{ multiple: true }}     | ${'box-shadow: 0 4px #4d5256,0 -4px #4d5256,4px 0 #4d5256,-4px 0 #4d5256; outline-color: #212529;'}
  ${'Select disabled'} | ${{ status: 'disabled' }} | ${'box-shadow: 0 4px #e9c46a,0 -4px #e9c46a,4px 0 #e9c46a,-4px 0 #e9c46a; outline-color: #b38106;'}
  ${'Select error'}    | ${{ status: 'error' }}    | ${'box-shadow: 0 4px #e76e55,0 -4px #e76e55,4px 0 #e76e55,-4px 0 #e76e55; outline-color: #ce372b;'}
  ${'Select success'}  | ${{ status: 'success' }}  | ${'box-shadow: 0 4px #92cc41,0 -4px #92cc41,4px 0 #92cc41,-4px 0 #92cc41; outline-color: #76c442;'}
`(`$displayName`, ({ props, css }) => {
  it(`renders with right styles`, () => {
    const { getByTestId } = render(
      <ThemeWrapper>
        <Select {...getProps(props)} />
      </ThemeWrapper>
    )
    const testedElement = getByTestId('select')
    expect(testedElement).toBeInTheDocument()

    expect(testedElement).toHaveStyle(css)
    // We cannot grab ::after pseudo element with testing-library.
    // Need to check if the dropdown arrow has color: transparent; on multiple
    // and color: [status]; for the rest.
    expect(testedElement.parentElement).toMatchSnapshot()
  })
})
