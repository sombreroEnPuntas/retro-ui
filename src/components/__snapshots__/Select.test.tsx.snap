// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select disabled renders with right styles 1`] = `
.c0 {
  font-family: 'Press Start 2P',monospace;
}

.c0::after {
  box-shadow: 3px 3px,6px 3px,9px 3px,12px 3px,15px 3px,6px 6px,9px 6px, 12px 6px,9px 9px;
  color: #b38106;
  content: '';
  display: block;
  height: 3px;
  pointer-events: none;
  position: relative;
  top: calc(-2.2 * 1.5em);
  left: calc(100% - 23px);
  width: 3px;
}

.c1 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: #ffe5a6;
  border: none;
  font-family: 'Press Start 2P',monospace;
  font-size: 1.5em;
  line-height: calc(2 * 1.5em);
  outline-color: #b38106;
  overflow: hidden;
  padding: 0.5rem 1rem;
  text-overflow: ellipsis;
  width: -webkit-fill-available;
  width: -moz-available;
  width: stretch;
  box-shadow: 0 4px #e9c46a,0 -4px #e9c46a,4px 0 #e9c46a,-4px 0 #e9c46a;
  margin: 4px;
}

.c1::-ms-expand {
  display: none;
}

.c1 option {
  background-color: white;
  color: #212529;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c1 option:checked {
  background-color: papayawhip;
  color: #212529;
}

@media (min-width:768px) {
  .c0::after {
    top: calc(-1.5 * 1.5em);
  }
}

@media (min-width:768px) {
  .c1 {
    line-height: 1.5em;
  }
}

<label
  class="c0"
  type="warning"
>
  test
  <select
    class="c1"
    data-testid="select"
    disabled=""
    name="test"
  >
    <option>
      &lt;
    </option>
    <option>
      X
    </option>
    <option>
      &gt;
    </option>
  </select>
</label>
`;

exports[`Select error renders with right styles 1`] = `
.c0 {
  font-family: 'Press Start 2P',monospace;
}

.c0::after {
  box-shadow: 3px 3px,6px 3px,9px 3px,12px 3px,15px 3px,6px 6px,9px 6px, 12px 6px,9px 9px;
  color: #ce372b;
  content: '';
  display: block;
  height: 3px;
  pointer-events: none;
  position: relative;
  top: calc(-2.2 * 1.5em);
  left: calc(100% - 23px);
  width: 3px;
}

.c1 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: white;
  border: none;
  font-family: 'Press Start 2P',monospace;
  font-size: 1.5em;
  line-height: calc(2 * 1.5em);
  outline-color: #ce372b;
  overflow: hidden;
  padding: 0.5rem 1rem;
  text-overflow: ellipsis;
  width: -webkit-fill-available;
  width: -moz-available;
  width: stretch;
  box-shadow: 0 4px #e76e55,0 -4px #e76e55,4px 0 #e76e55,-4px 0 #e76e55;
  margin: 4px;
}

.c1::-ms-expand {
  display: none;
}

.c1 option {
  background-color: white;
  color: #212529;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c1 option:checked {
  background-color: papayawhip;
  color: #212529;
}

@media (min-width:768px) {
  .c0::after {
    top: calc(-1.5 * 1.5em);
  }
}

@media (min-width:768px) {
  .c1 {
    line-height: 1.5em;
  }
}

<label
  class="c0"
  type="error"
>
  test
  <select
    class="c1"
    data-testid="select"
    name="test"
  >
    <option>
      &lt;
    </option>
    <option>
      X
    </option>
    <option>
      &gt;
    </option>
  </select>
</label>
`;

exports[`Select multiple renders with right styles 1`] = `
.c0 {
  font-family: 'Press Start 2P',monospace;
}

.c0::after {
  box-shadow: 3px 3px,6px 3px,9px 3px,12px 3px,15px 3px,6px 6px,9px 6px, 12px 6px,9px 9px;
  color: transparent;
  content: '';
  display: block;
  height: 3px;
  pointer-events: none;
  position: relative;
  top: calc(-2.2 * 1.5em);
  left: calc(100% - 23px);
  width: 3px;
}

.c1 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: white;
  border: none;
  font-family: 'Press Start 2P',monospace;
  font-size: 1.5em;
  line-height: calc(2 * 1.5em);
  outline-color: #212529;
  overflow: hidden;
  padding: 0.5rem 1rem;
  text-overflow: ellipsis;
  width: -webkit-fill-available;
  width: -moz-available;
  width: stretch;
  box-shadow: 0 4px #4d5256,0 -4px #4d5256,4px 0 #4d5256,-4px 0 #4d5256;
  margin: 4px;
}

.c1::-ms-expand {
  display: none;
}

.c1 option {
  background-color: white;
  color: #212529;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c1 option:checked {
  background-color: papayawhip;
  color: #212529;
}

@media (min-width:768px) {
  .c0::after {
    top: calc(-1.5 * 1.5em);
  }
}

@media (min-width:768px) {
  .c1 {
    line-height: 1.5em;
  }
}

<label
  class="c0"
>
  test
  <select
    class="c1"
    data-testid="select"
    multiple=""
    name="test"
  >
    <option>
      &lt;
    </option>
    <option>
      X
    </option>
    <option>
      &gt;
    </option>
  </select>
</label>
`;

exports[`Select renders with right styles 1`] = `
.c0 {
  font-family: 'Press Start 2P',monospace;
}

.c0::after {
  box-shadow: 3px 3px,6px 3px,9px 3px,12px 3px,15px 3px,6px 6px,9px 6px, 12px 6px,9px 9px;
  color: #212529;
  content: '';
  display: block;
  height: 3px;
  pointer-events: none;
  position: relative;
  top: calc(-2.2 * 1.5em);
  left: calc(100% - 23px);
  width: 3px;
}

.c1 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: white;
  border: none;
  font-family: 'Press Start 2P',monospace;
  font-size: 1.5em;
  line-height: calc(2 * 1.5em);
  outline-color: #212529;
  overflow: hidden;
  padding: 0.5rem 1rem;
  text-overflow: ellipsis;
  width: -webkit-fill-available;
  width: -moz-available;
  width: stretch;
  box-shadow: 0 4px #4d5256,0 -4px #4d5256,4px 0 #4d5256,-4px 0 #4d5256;
  margin: 4px;
}

.c1::-ms-expand {
  display: none;
}

.c1 option {
  background-color: white;
  color: #212529;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c1 option:checked {
  background-color: papayawhip;
  color: #212529;
}

@media (min-width:768px) {
  .c0::after {
    top: calc(-1.5 * 1.5em);
  }
}

@media (min-width:768px) {
  .c1 {
    line-height: 1.5em;
  }
}

<label
  class="c0"
>
  test
  <select
    class="c1"
    data-testid="select"
    name="test"
  >
    <option>
      &lt;
    </option>
    <option>
      X
    </option>
    <option>
      &gt;
    </option>
  </select>
</label>
`;

exports[`Select success renders with right styles 1`] = `
.c0 {
  font-family: 'Press Start 2P',monospace;
}

.c0::after {
  box-shadow: 3px 3px,6px 3px,9px 3px,12px 3px,15px 3px,6px 6px,9px 6px, 12px 6px,9px 9px;
  color: #76c442;
  content: '';
  display: block;
  height: 3px;
  pointer-events: none;
  position: relative;
  top: calc(-2.2 * 1.5em);
  left: calc(100% - 23px);
  width: 3px;
}

.c1 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: white;
  border: none;
  font-family: 'Press Start 2P',monospace;
  font-size: 1.5em;
  line-height: calc(2 * 1.5em);
  outline-color: #76c442;
  overflow: hidden;
  padding: 0.5rem 1rem;
  text-overflow: ellipsis;
  width: -webkit-fill-available;
  width: -moz-available;
  width: stretch;
  box-shadow: 0 4px #92cc41,0 -4px #92cc41,4px 0 #92cc41,-4px 0 #92cc41;
  margin: 4px;
}

.c1::-ms-expand {
  display: none;
}

.c1 option {
  background-color: white;
  color: #212529;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c1 option:checked {
  background-color: papayawhip;
  color: #212529;
}

@media (min-width:768px) {
  .c0::after {
    top: calc(-1.5 * 1.5em);
  }
}

@media (min-width:768px) {
  .c1 {
    line-height: 1.5em;
  }
}

<label
  class="c0"
  type="success"
>
  test
  <select
    class="c1"
    data-testid="select"
    name="test"
  >
    <option>
      &lt;
    </option>
    <option>
      X
    </option>
    <option>
      &gt;
    </option>
  </select>
</label>
`;
